src/chainstate/stacks/index/cache.rs

```diff
@@ -0,0 +1,880 @@
+            block_id_cache: HashMap::new(),
+            node_cache: HashMap::new(),
+            hash_cache: HashMap::new(),
+        }
+    }
+
+    /// Obtain a possibly-cached node and its hash.
+    /// Only return data if we have *both* the node and hash
+    pub fn load_node_and_hash(
+        &self,
+        block_id: u32,
+        trieptr: &TriePtr,
+    ) -> Option<(TrieNodeType, TrieHash)> {
+        match (
+            self.load_node(block_id, trieptr),
+            self.load_node_hash(block_id, trieptr),
+        ) {
+            (Some(ref node), Some(ref hash)) => Some(((*node).clone(), (*hash).clone())),
+            _ => None,
+        }
+    }
+
+    /// Obtain a possibly-cached node
+    pub fn load_node(&self, block_id: u32, trieptr: &TriePtr) -> Option<TrieNodeType> {
+        match self
+            .node_cache
+            .get(&TrieNodeAddr(block_id, trieptr.clone()))
+        {
+            Some(ref node) => Some((*node).clone()),
+            None => None,
+        }
+    }
+
+    /// Obtain a possibly-cached node hash
+    pub fn load_node_hash(&self, block_id: u32, trieptr: &TriePtr) -> Option<TrieHash> {
+        self.hash_cache
+            .get(&TrieNodeAddr(block_id, trieptr.clone()))
+            .cloned()
+    }
+
+    /// Cache a node and hash
+    pub fn store_node_and_hash(
+        &mut self,
+        block_id: u32,
+        trieptr: TriePtr,
+        node: TrieNodeType,
+        hash: TrieHash,
+    ) {
+        self.store_node(block_id, trieptr.clone(), node);
+        self.store_node_hash(block_id, trieptr, hash)
+    }
+
+    /// Cache just a node
+    pub fn store_node(&mut self, block_id: u32, trieptr: TriePtr, node: TrieNodeType) {
+        self.node_cache
+            .insert(TrieNodeAddr(block_id, trieptr.clone()), node);
+    }
+
+    /// Cache just a node hash
+    pub fn store_node_hash(&mut self, block_id: u32, trieptr: TriePtr, hash: TrieHash) {
+        self.hash_cache
+            .insert(TrieNodeAddr(block_id, trieptr), hash);
+    }
+
+    /// Load up a block hash, given its ID
+    pub fn load_block_hash(&self, block_id: u32) -> Option<T> {
+        self.block_hash_cache.get(&block_id).cloned()
+    }
+
+    /// Cache a block hash, given its ID
+    pub fn store_block_hash(&mut self, block_id: u32, block_hash: T) {
+        assert!(!self.block_hash_cache.contains_key(&block_id));
+        self.block_id_cache.insert(block_hash.clone(), block_id);
+        self.block_hash_cache.insert(block_id, block_hash);
+    }
+
+    /// Get an immutable reference to a block hash, given the ID
+    pub fn ref_block_hash(&self, block_id: u32) -> Option<&T> {
+        self.block_hash_cache.get(&block_id)
+    }
+
+    /// Get the block ID, given its hash
+    pub fn load_block_id(&self, block_hash: &T) -> Option<u32> {
+        self.block_id_cache.get(block_hash).map(|id| *id)
+    }
+}
+
+/// Trie node cache strategies
+pub enum TrieCache<T: MarfTrieId> {
+    /// Do nothing
+    Noop(TrieCacheState<T>),
```

Comment:
"Nothing"?

Reply:
Yup. Don't do anything at all.  It's a no-op strategy.

Reply:
Probably more descriptive as `/// Cache nothing`